specify/specify7
# Add new age filter in QB

Fixes #5180

### Checklist

      and self-explanatory (or properly documented)

### Testing instructions
This PR affects database migrations. See migration testing instructions ([here](https://specify.github.io/testing/pull_requests.html))

https://ciscollectionsageqbmigration52722-issue-5180.test.specifysystems.org/specify/query/new/collectionobject/

Use the `ciscollections_age_qb` database in the test panel.  Use the 'Rocks' collection.

Go to the QB and select Collection Object, then verify the following queries.  Note the the value returned in the 'Age' column is the CollectionObjectID, so no need to add an ID column in the query.

![image](https://github.com/user-attachments/assets/77cb0449-0882-4c34-aaea-7f88ae6eb882)

- [ ] Create a Collection Object Query.  Select the 'Age' field.  Select the 'Range' filter.  Try setting the 'start_time' to a value that is less than the 'end_time' value.  When the query runs, it should through an error "Start time must be greater than or equal to end time".
- [ ] Create a Collection Object Query.  Select the 'Age' field.  Select the 'Any' filter.  Run the query and make sure no errors occur.
- [ ] Create a Collection Object Query.  Select the 'Age' field.  Select the 'Range' filter.  Use the following table to make multiple queries with differing values for the 'start_time', 'end_time', and 'strict' in the 'Range' filter.  Make sure that for all these queries, that the value `40057` is present in the query results of the 'Age' column.

![image](https://github.com/user-attachments/assets/15ae4dd7-f6e4-4919-a524-49d52f38f1c2)
![image](https://github.com/user-attachments/assets/9a104096-50f7-43b3-aa9d-4d92478888a6)

| start_time | end_time | strict |
| ---------- | -------- | ------ |
| 300        | 100      | false  |
| 300        | 200      | false  |
| 200        | 100      | false  |
| 300        | 201      | false  |
| 199        | 100      | false  |
| 300        | 199      | false  |
| 201        | 100      | false  |
| 300        | 100      | true   |
| 201        | 199      | true   |

- [ ] Create a Collection Object Query.  Select the 'Age' field.  Select the 'Range' filter.  Use the following table to make multiple queries with differing values for the 'start_time', 'end_time', and 'strict' in the 'Range' filter.  Make sure that for all these queries, that the value `40057` is __NOT__ present in the query results of the 'Age' column.

| start_time | end_time | strict |
| ---------- | -------- | ------ |
| 300        | 250      | false  |
| 150        | 100      | false  |
| 300        | 250      | true   |
| 150        | 100      | true   |

- [ ] Create a Collection Object Query.  Select the 'Age' field.  Select the 'Range' filter.  Use the following table to make multiple queries with differing values for the 'start_time', 'end_time', and 'strict' in the 'Range' filter.  Make sure that for all these queries, that the value `40058` is present in the query results of the 'Age' column.

| start_time | end_time | strict |
| ---------- | -------- | ------ |
| 220        | 140      | false  |
| 180        | 160      | false  |
| 220        | 100      | false  |
| 300        | 140      | false  |
| 180        | 100      | false  |
| 300        | 160      | false  |
| 300        | 100      | false  |
| 300        | 100      | true   |
| 195        | 145      | true   |

- [ ] Create a Collection Object Query.  Select the 'Age' field.  Select the 'Range' filter.  Use the following table to make multiple queries with differing values for the 'start_time', 'end_time', and 'strict' in the 'Range' filter.  Make sure that for all these queries, that the value `40058` is __NOT__ present in the query results of the 'Age' column.

| start_time | end_time | strict |
| ---------- | -------- | ------ |
| 300        | 250      | false  |
| 120        | 100      | false  |
| 300        | 250      | true   |
| 120        | 100      | true   |
| 180        | 160      | true   |
| 180        | 100      | true   |
| 300        | 160      | true   |

- [ ] Create a Collection Object Query.  Select the 'Age' field.  Select the 'Range' filter.  Use the following table to make multiple queries with differing values for the 'start_time', 'end_time', and 'strict' in the 'Range' filter.  Make sure that for all these queries, that the value `40059` is present in the query results of the 'Age' column.

| start_time | end_time | strict |
| ---------- | -------- | ------ |
| 200        | 100      | false  |
| 200        | 10       | false  |

- [ ] Create a Collection Object Query.  Select the 'Age' field.  Select the 'Range' filter. Set the 'start_time' to 170 and 'end_time' to 130.  Run the query and verify that the Collection Object `40060` is returned for a query results.
- [ ] Create a Collection Object Query.  Select the 'Age' field.  Select the 'Range' filter. Set the 'start_time' to 550 and 'end_time' to 500.  Run the query and verify that the Collection Object `40061` is returned for a query results.
- [ ] Create a Collection Object Query.  Select the 'Age' field.  Select the 'Range' filter. Set the 'start_time' to 200 and 'end_time' to 140.  Run the query and verify that the Collection Object `40063` is returned for a query results.
- [ ] Test the age query with the 'Name'.  Here are a defined subset of CollectionObjectIDs that should be included in the returned query results with a given age period name:

| name          | co_id |
| ------------- | ----- |
| Oxfordian     | 40063 |
| Jurassic      | 40060 |
| Jurassic      | 40058 |

- [ ] Do age queries with 'Any' and explore including the follow query field paths and make sure that column in the query results have at least one row with data populated.  Might be good to just set each field to 'Not Empty' one at a time, to make sure rows are returned:
	- [ ] collectionobject->absoluteage
	- [ ] ollectionobject->relativeage->chronostrat
	- [ ] collectionobject->paleocontext->chronostrat
	- [ ] collectionobject->collectionevent->paleocontext->chronostrat
	- [ ] Collectionobject->collectionevent->locality->paleocontext->chronostrat

![image](https://github.com/user-attachments/assets/0adf1121-4fc6-419d-ae25-37e890f73057)
